File: backend/slots.py
"""Timeline slots: a cut made of positions, not of file paths.

Contract §7.1 is the whole design: *a timeline visual clip is fundamentally a
slot tied to a DirectorShot, not merely a file path.*

    Timeline Slot -> Shot 3C -> Selected Output Take B

The difference shows the moment anything changes. When a cut is a list of file
paths, choosing a different take means rebuilding the list, and everything the
editor did to those clips — trims, order, what sits either side — is rebuilt
with it, which is to say lost. When a cut is a list of slots, a new take is a
new *value* in a slot that already exists, and the edit survives because the
edit was never made of files.

The same reasoning covers placeholders (§6.2). A missing shot is not an absence,
it is a slot whose media has not arrived: it holds the shot it belongs to, the
duration it is owed, the media type it expects, and the beat it came from. That
is why a later output can drop into it without reconstructing the edit — the
placeholder was already the edit. A placeholder that is only a grey rectangle on
screen looks the same and carries none of that, which is the failure this model
exists to prevent.

What a slot deliberately does NOT own: the creative intent (Director's), the
attempts that produced its media (Generate's), or approval. It references them.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from pathlib import Path

from . import atomic, config

VIDEO = "video"
STILL = "still"


@dataclass
class TimelineSlot:
    """One visual position in the cut."""

    id: str                       # stable across take swaps and re-plans
    beat_id: str                  # the narration beat this covers
    shot_id: str = ""             # the DirectorShot; "" for a whole-beat clip
    index: int = 0                # position in the cut
    intended_duration: float = 0.0
    expected_media: str = VIDEO
    media: str = ""               # media-root-relative; "" while a placeholder
    source_attempt: str = ""      # the GenerationAttempt that produced `media`
    trim_in: float = 0.0
    trim_out: float = 0.0         # 0 means "to the end of the media"
    note: str = ""

    @property
    def placeholder(self) -> bool:
        """A slot is a placeholder exactly when it has no media yet.

        Derived, not stored, so the two can never disagree — a stored flag left
        true after media arrived is precisely how a filled slot goes on
        rendering as a placeholder.
        """
        return not self.media

    @property
    def duration(self) -> float:
        """What this slot occupies in the cut.

        The intended duration governs until a trim says otherwise. A placeholder
        therefore holds its full length, which is what keeps everything after it
        in the right place while its media is still missing.
        """
        if self.trim_out and self.trim_out > self.trim_in:
            return round(self.trim_out - self.trim_in, 3)
        return round(max(0.0, self.intended_duration - self.trim_in), 3)

    def to_dict(self) -> dict:
        d = asdict(self)
        d["placeholder"] = self.placeholder
        d["duration"] = self.duration
        return d


# --- identity -------------------------------------------------------------------

def slot_id(beat_id: str, shot_id: str = "") -> str:
    """Stable slot identity, derived from what the slot covers.

    Derived rather than random so a slot rebuilt from the same plan is the SAME
    slot and keeps its trims. Random ids would make every rebuild a new cut.
    """
    return f"{beat_id}::{shot_id}" if shot_id else f"{beat_id}::beat"


# --- storage --------------------------------------------------------------------

def slots_path() -> Path:
    return config.project_dir() / "timeline_slots.json"


def load() -> list[TimelineSlot]:
    p = slots_path()
    if not p.is_file():
        return []
    try:
        raw = atomic.read_json(p)
    except (OSError, ValueError):
        # A cut that cannot be read is not an empty cut. Callers rebuild from
        # the plan instead, which loses trims but never silently ships a film
        # with everything after the unreadable point in the wrong place.
        raise
    known = set(TimelineSlot.__dataclass_fields__)
    return [TimelineSlot(**{k: v for k, v in s.items() if k in known})
            for s in (raw.get("slots") or [])]


def save(slots: list[TimelineSlot]) -> Path:
    return atomic.write_json(slots_path(),
                             {"slots": [asdict(s) for s in slots]})


# --- building and reconciling -----------------------------------------------------

def build(sb, coverage_for=None, media_for=None) -> list[TimelineSlot]:
    """The slots a storyboard implies, in cut order.

    ``coverage_for(beat_id)`` returns that beat's DirectorShots, or None when
    the beat has no coverage plan and is therefore one whole-beat slot.
    ``media_for(beat_id, shot_id)`` returns ``(media, attempt_id)`` or None.

    Both are injected so this module never reaches into the director's or the
    generation ledger's storage, and so a caller that already has plans loaded
    does not pay to read them again.
    """
    out: list[TimelineSlot] = []
    for beat in sb.shots:
        shots = (coverage_for(beat.scene_id) if coverage_for else None) or []
        if shots:
            for ds in shots:
                media, attempt = (media_for(beat.scene_id, ds.id)
                                  if media_for else (None, None)) or ("", "")
                out.append(TimelineSlot(
                    id=slot_id(beat.scene_id, ds.id),
                    beat_id=beat.scene_id, shot_id=ds.id, index=len(out),
                    intended_duration=round(float(ds.duration), 3),
                    expected_media=STILL if ds.motion_type == "static" else VIDEO,
                    media=media or "", source_attempt=attempt or "",
                ))
        else:
            media, attempt = (media_for(beat.scene_id, "")
                              if media_for else (None, None)) or ("", "")
            dur = float(beat.camera.duration) if beat.camera else 0.0
            out.append(TimelineSlot(
                id=slot_id(beat.scene_id), beat_id=beat.scene_id, index=len(out),
                intended_duration=round(dur, 3), expected_media=VIDEO,
                media=media or "", source_attempt=attempt or "",
            ))
    return out


def reconcile(existing: list[TimelineSlot],
              rebuilt: list[TimelineSlot]) -> list[TimelineSlot]:
    """Fold a rebuilt slot list onto the edit that already exists.

    Slots that still exist keep their trims and their note — that is the editing
    work, and a re-plan of one beat must not discard it everywhere. Slots whose
    shot no longer exists are dropped; new ones arrive empty.

    A trim is kept only while it still fits the slot's intended duration. When
    Director shortens a shot below an existing trim, the trim is dropped rather
    than silently retimed, because quietly reinterpreting an editor's decision
    is worse than making them redo it: §7.1 forbids hiding a plan change by
    stretching the cut around it.
    """
    prior = {s.id: s for s in existing}
    out: list[TimelineSlot] = []
    for i, fresh in enumerate(rebuilt):
        old = prior.get(fresh.id)
        if old is not None:
            fresh.note = old.note
            fits = (old.trim_in < fresh.intended_duration and
                    (not old.trim_out or old.trim_out <= fresh.intended_duration))
            if fits:
                fresh.trim_in, fresh.trim_out = old.trim_in, old.trim_out
            else:
                fresh.note = (old.note + " " if old.note else "") + (
                    f"trim dropped: the shot is now {fresh.intended_duration:.2f}s")
            # An existing slot keeps media the caller did not supply, so a
            # rebuild that cannot see the ledger does not blank the cut.
            if not fresh.media and old.media:
                fresh.media, fresh.source_attempt = old.media, old.source_attempt
        fresh.index = i
        out.append(fresh)
    return out


def fill(slots: list[TimelineSlot], shot_id: str, media: str,
         attempt: str = "", beat_id: str = "") -> TimelineSlot | None:
    """Put media into the slot that already exists. §7.1's core operation.

    Selecting a different take must not recreate the edit, so this changes one
    field and touches nothing else: the slot keeps its id, its index, its trims
    and everything around it.
    """
    for s in slots:
        if s.shot_id == shot_id and (not beat_id or s.beat_id == beat_id):
            s.media = media
            s.source_attempt = attempt
            return s
    return None


def coverage(slots: list[TimelineSlot]) -> dict:
    """How complete the cut is, in the words §6.2 asks for."""
    total = len(slots)
    ready = sum(1 for s in slots if not s.placeholder)
    missing = total - ready
    return {
        "slots": total,
        "ready": ready,
        "placeholders": missing,
        "runtime": round(sum(s.duration for s in slots), 3),
        "summary": (f"{ready}/{total} visuals ready"
                    + (f" · Draft 1 will use {missing} placeholder"
                       f"{'' if missing == 1 else 's'}" if missing else "")),
    }

File: frontend/src/lib/slots.ts
/**
 * The cut, as the server describes it. Client-side mirror of `backend/slots.py`.
 *
 * Contract §7.1: a timeline visual clip is a slot tied to a DirectorShot, not a
 * file path. Everything here is a *reader* of `GET /api/timeline/slots` — the
 * client never derives slot state it was not told, because two of the fields the
 * server sends are derived there precisely so the two sides cannot disagree:
 *
 *   - `placeholder` is `not media`, computed per request. A stored flag left
 *     true after media arrived is how a filled slot goes on rendering as a
 *     placeholder; a client flag mirroring it is the same bug with a longer
 *     fuse. §11.4.
 *   - `coverage.summary` is the §6.2 sentence, written server-side. A client
 *     that counts for itself will eventually disagree with the server, and the
 *     UI would then be stating something no authoritative state supports.
 *
 * So this module computes exactly one thing the server does not: horizontal
 * positions, which are a property of the screen and of nothing else.
 */

/** One visual position in the cut. Field-for-field `TimelineSlot.to_dict()`. */
export interface TimelineSlot {
  /** Stable identity, `beat_id::shot_id` — derived, not random, so a slot
   *  rebuilt from the same plan is the SAME slot and keeps its trims. */
  id: string;
  /** The narration beat this covers. C5's "source beat". */
  beat_id: string;
  /** The DirectorShot; "" for a whole-beat clip. */
  shot_id: string;
  index: number;
  intended_duration: number;
  /** "video" | "still" — C5's "expected media type". */
  expected_media: string;
  /** Media-root-relative; "" while a placeholder. */
  media: string;
  /** The GenerationAttempt that produced `media`. */
  source_attempt: string;
  trim_in: number;
  /** 0 means "to the end of the media". */
  trim_out: number;
  note?: string;
  /** DERIVED SERVER-SIDE. Never recompute, never mirror in client state. */
  placeholder?: boolean;
  /** DERIVED SERVER-SIDE: trim if set, else intended_duration - trim_in. */
  duration?: number;
}

/** `coverage()` — how complete the cut is, in the words §6.2 asks for. */
export interface SlotCoverage {
  slots: number;
  ready: number;
  placeholders: number;
  runtime: number;
  /** e.g. "3/5 visuals ready · Draft 1 will use 2 placeholders". Render it. */
  summary: string;
}

export interface SlotBlock {
  slot: TimelineSlot;
  start: number;
  dur: number;
}

/** The takes the server holds for one slot's DirectorShot, as it reports them. */
export interface SlotTakes {
  variations: string[];
  /** 0-based, as everywhere on the server; null when nothing is chosen. */
  chosen: number | null;
}

export const STILL = "still";

/**
 * Whether the server says this slot has media in it.
 *
 * Fails closed on purpose: filled only when the server explicitly said
 * `placeholder: false`. An older or partial payload that omits the field is
 * therefore drawn as a placeholder rather than as a clip we cannot vouch for,
 * which is what §11.4 asks — the UI may not claim a slot is filled unless
 * authoritative backend state agrees.
 *
 * Deliberately does NOT look at `slot.media`, and must not: re-deriving the flag
 * client-side is how the two come to disagree.
 */
export function isFilled(slot: TimelineSlot): boolean {
  return slot.placeholder === false;
}

/** What a slot occupies in the cut. Server-derived; `duration` is authoritative.
 *
 * The fallback is `TimelineSlot.duration`'s arithmetic, clause for clause: a
 * trim governs when one is set, otherwise the intended length less trim_in.
 * Anything else would make this module disagree with the server inside a file
 * whose whole premise is that the two must not.
 */
export function slotDuration(slot: TimelineSlot): number {
  if (typeof slot.duration === "number") return slot.duration;
  if (slot.trim_out && slot.trim_out > slot.trim_in) {
    return Math.round((slot.trim_out - slot.trim_in) * 1000) / 1000;
  }
  return Math.round(Math.max(0, slot.intended_duration - slot.trim_in) * 1000) / 1000;
}

/** True when an editor has trimmed this slot away from its intended length. */
export function isTrimmed(slot: TimelineSlot): boolean {
  return slot.trim_in > 0 || slot.trim_out > 0;
}

/** How a slot names itself: the DirectorShot, or the beat it covers whole. */
export function slotLabel(slot: TimelineSlot): string {
  return slot.shot_id || `${slot.beat_id} · whole beat`;
}

/**
 * What a placeholder is waiting for, in words (C5).
 *
 * A placeholder that is only a grey rectangle carries none of what the slot
 * knows, so a user cannot tell what is missing. This sentence is the difference.
 */
export function waitingFor(slot: TimelineSlot): string {
  const kind = slot.expected_media === STILL ? "a still" : "a video clip";
  const owner = slot.shot_id ? `shot ${slot.shot_id}` : `beat ${slot.beat_id}`;
  return `waiting for ${kind} for ${owner}`;
}

/**
 * Where each slot sits on screen, in seconds.
 *
 * Slots are laid out inside their beat: `beatStarts` anchors the first slot of
 * each beat to where that beat already begins on the audio tracks, and the rest
 * follow it. Without the anchor the video track would accumulate its own
 * timebase and drift away from narration the moment a trim shortened a slot —
 * the cut would look retimed when only one clip had been shortened, which is
 * the silent stretch §7.1 forbids.
 *
 * Slots are read in `index` order: `index` is the position in the cut and is
 * the server's, not the array's.
 */
export function layoutSlots(
  slots: TimelineSlot[],
  beatStarts: Record<string, number> = {},
): { blocks: SlotBlock[]; total: number } {
  const ordered = [...slots].sort((a, b) => a.index - b.index);
  const blocks: SlotBlock[] = [];
  let acc = 0;
  let beat: string | null = null;
  for (const slot of ordered) {
    if (slot.beat_id !== beat) {
      beat = slot.beat_id;
      const anchored = beatStarts[slot.beat_id];
      if (typeof anchored === "number") acc = anchored;
    }
    const dur = slotDuration(slot);
    blocks.push({ slot, start: acc, dur });
    acc += dur;
  }
  return { blocks, total: acc };
}

// --- whose cut is this? ----------------------------------------------------

/**
 * Everything the studio holds about one project's cut, stamped with whose it is.
 *
 * Slot ids are `beat_id::shot_id`, which is unique within a film and emphatically
 * not across films. Held unstamped, a slot's takes from film A are served for a
 * slot in film B whose beat and shot ids happen to coincide — and beat ids are
 * `s001`, `s002`, so they coincide constantly. The page already refuses replies
 * that arrive for the wrong project (`isStaleReply`); this is the same rule for
 * state that has already arrived and is merely still on screen.
 */
export interface SlotView {
  /** The project this was read for. Never inferred; set when it was read. */
  projectId: string | null;
  slots: TimelineSlot[] | null;
  coverage: SlotCoverage | null;
  error: string | null;
  takes: Record<string, SlotTakes>;
}

/** No cut read yet — which is not the same claim as a cut with no clips. */
export const NO_SLOT_VIEW: SlotView = {
  projectId: null, slots: null, coverage: null, error: null, takes: {},
};

/**
 * The held cut, but only if it belongs to the project on screen.
 *
 * Returns the unread view rather than someone else's film. The timeline draws
 * `slots: null` as "reading the cut…", which is the truthful thing to say in the
 * moment between switching project and the new cut arriving.
 */
export function viewForProject(held: SlotView, projectId: string | null): SlotView {
  return held.projectId === projectId ? held : NO_SLOT_VIEW;
}
